### PR TITLE
Wirelink now applies input serializers to "zero" values if present.

### LINKED ISSUE #1041: Unwired Wirelink can cause Overwriting of Default Table
## Observed Behavior:

The following E2 code, when the "Unwired" input is left, well, unwired, causes the default E2 table to become writable. This then affects any and all new persisted tables that are not overwritten with "T = table()".
#### "Infection" Code:

```
@inputs Unwired:wirelink

T = Unwired["AnyNameHere",table]
T["This Value Will Appear In All New Persisted Tables",number] = 1337
```
#### Display Code:

```
@persist Test:table

#Test = table()     #If this is uncommented, the issue does not occur.
                    #Also, this issue does not occur with non-persisted tables.
printTable(Test)
```
## Steps to Reproduce:
1. Start a new online session.
2. Open the E2 tool and paste the _"Infection" Code_ E2, then place into the world.
3. Repeat Step 2 with the _Display Code_ E2.
4. Observe that the _Display Code_ E2 outputs _"This Value Will Appear In All New Persisted Tables  =   1337"_
## Notes:
1. This issue occurred when only Wiremod was installed.
2. Any value of any name and type can be written to the default table when using the above code.
3. This issue persists even after the _"Infection Code"_ chip has been removed, and is only resolved after running "wire_expression2_reload".
4. This issue does not occur when setting the persisted table with "Table = table()", or with local Tables.

### PR DESCRIPTION
**Pull request by TomyLobo**
Can someone test this? :)

Fixes #1041

### CODE REVIEW

**Comment by Divran** [comment]
won't work for tables because the table input serializer doesn't copy. You need to use table.Copy of the zero value is a lua table

**Comment by TomyLobo** [comment]
table's input serializer does copy DEFAULT

**Comment by TomyLobo** [comment]
There, how about that?
Why do you replace empty tables by a copy of DEFAULT anyway?
I assume stuff outside of E2, but if you account for that, why not for non-empty tables that lack the n, ntypes etc fields?

Also, if i output a default table through any means and reinput it to an E2, it's still the same table. That's not good either and my last commit fixes that as well.

**Comment by TomyLobo** [comment]
Btw, where's the table(...) function and what's kvtable?
I'm having a bad feeling about this...

**Comment by Divran** [comment]
input == DEFAULT won't work because that only checks if the table is the exact same one, not if it just has the same contents

**Comment by TomyLobo** [comment]
As @Steeveeoo can probably confirm, the problem IS that the table is the exact same one as DEFAULT in every case.

As for third-party addons: They have to take their own precautions. They can grab DEFAULT from wire_expression2_types, just like wirelink.lua does.

table(...): found it :)

**Comment by AbigailBuccaneer** [comment]
Looks good to me - feel free to merge once the conflicts are resolved.

**Comment by AbigailBuccaneer** [comment]
@TomyLobo this doesn't fix behaviour seen in the example code in #1041.

**Comment by TomyLobo** [comment]
crap... will take a closer look once i get home

**Comment by TomyLobo** [comment]
I pushed the two general cleanup commits onto master and rebased the rest to fix the conflicts.
@AbigailBuccaneer what exactly happens when you try it?

**Comment by TomyLobo** [comment]
I just noticed that my patch does nothing for types whose zero element is an empty table.
Is that still the case for the table type?
Either way, it's a necessary fix nonetheless, since it prevents the same issue from happening for types other than table/array, once we determine and fix the root cause..

**Comment by TomyLobo** [comment]
Ok so this problem got introduced when @Divran replaced table.lua with mtable.lua all the way back in 2010:
https://github.com/wiremod/wire/commit/77e1deb1cbc2014febb3960b200a5cdf37a2032e#diff-8eb7409a45bbe9ed5826e029ca7a56ddL7

**Comment by Divran** [comment]
Yes it appears I didn't know of this issue back then.

**Comment by Divran** [comment]
Any updates in this pr?

**Comment by TomyLobo** [comment]
@Divran I'm still waiting for @AbigailBuccaneer to tell me what happens when he tries that sample code.

**Comment by TomyLobo** [comment]
@Divran @AbigailBuccaneer what happened to this?

**Comment by TomyLobo** [comment]
> AbigailBuccaneer was unassigned by TomyLobo 26 minutes ago

wat, i did no such thing... I guess it's because I made a modification to the PR (comment) while @AbigailBuccaneer did not have push rights.

**Comment by Divran** [comment]
@TomyLobo is this alright to merge now?

**Comment by AbigailBuccaneer** [comment]
Will test tomorrow.

**Comment by Nebual** [comment]
Didn't seem to actually fix the issue, since IsEmpty(table) doesn't work - has to be `IsEmpty(table.n) and IsEmpty(table.s)` if I'm understanding correctly. Pushed a commit with that addition.

File: lua/entities/gmod_wire_expression2/core/table.lua [hunk 1]
```diff
@@ -37,7 +37,7 @@ local DEFAULT = {n={},ntypes={},s={},stypes={},size=0}
 
 registerType("table", "t", table.Copy(DEFAULT),
 	function(self, input)
-		if IsEmpty(input) then
+		if IsEmpty(input.n) and IsEmpty(input.s) then
```
Comment (AbigailBuccaneer) [inline lua/entities/gmod_wire_expression2/core/table.lua]: Would `input.size == 0` work? (I see `DEFAULT` has a `size`, but I don't know if it's actually used anywhere.)

Comment (Divran) [reply lua/entities/gmod_wire_expression2/core/table.lua]: yep `t.size` is used

**Pull request merged**